internal/storage/serde_events.go

```diff
@@ -0,0 +1,575 @@
+			return err
+		}
+	}
+
+	composeRecord := func() bool {
+		for i, rr := range crr.rrs {
+			if ok := rr.Next(); !ok {
+				return false
+			}
+			// compose record
+			crr.r.recs[crr.fields[i]] = rr.Record()
+		}
+		return true
+	}
+
+	// Try compose records
+	if ok := composeRecord(); !ok {
+		// If failed the first time, try iterate next batch (blob), the error may be io.EOF
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+		// If iterate next batch success, try compose again
+		if ok := composeRecord(); !ok {
+			// If the next blob is empty, return io.EOF (it's rare).
+			return io.EOF
+		}
+	}
+	return nil
+}
+
+func (crr *compositeBinlogRecordReader) Record() Record {
+	return &crr.r
+}
+
+func (crr *compositeBinlogRecordReader) Close() {
+	for _, close := range crr.closers {
+		if close != nil {
+			close()
+		}
+	}
+}
+
+func parseBlobKey(blobKey string) (colId FieldID, logId UniqueID) {
+	if blobKey == "" {
+		return 0, 0
+	}
+	if _, _, _, colId, logId, ok := metautil.ParseInsertLogPath(blobKey); ok {
+		return colId, logId
+	}
+	if colId, err := strconv.ParseInt(blobKey, 10, 64); err == nil {
+		// data_codec.go generate single field id as blob key.
+		return colId, 0
+	}
+	return -1, -1
+}
+
+func newCompositeBinlogRecordReader(blobs []*Blob) (*compositeBinlogRecordReader, error) {
+	sort.Slice(blobs, func(i, j int) bool {
+		iCol, iLog := parseBlobKey(blobs[i].Key)
+		jCol, jLog := parseBlobKey(blobs[j].Key)
+
+		if iCol == jCol {
+			return iLog < jLog
+		}
+		return iCol < jCol
+	})
+
+	blobm := make([][]*Blob, 0)
+	var fieldId FieldID = -1
+	var currentCol []*Blob
+
+	for _, blob := range blobs {
+		colId, _ := parseBlobKey(blob.Key)
+		if colId != fieldId {
+			if currentCol != nil {
+				blobm = append(blobm, currentCol)
+			}
+			currentCol = make([]*Blob, 0)
+			fieldId = colId
+		}
+		currentCol = append(currentCol, blob)
+	}
+	if currentCol != nil {
+		blobm = append(blobm, currentCol)
+	}
+
+	return &compositeBinlogRecordReader{
+		blobs: blobm,
+	}, nil
+}
+
+func NewBinlogDeserializeReader(blobs []*Blob, PKfieldID UniqueID) (*DeserializeReader[*Value], error) {
+	reader, err := newCompositeBinlogRecordReader(blobs)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewDeserializeReader(reader, func(r Record, v []*Value) error {
+		// Note: the return value `Value` is reused.
+		for i := 0; i < r.Len(); i++ {
+			value := v[i]
+			if value == nil {
+				value = &Value{}
+				m := make(map[FieldID]interface{}, len(r.Schema()))
+				value.Value = m
+				v[i] = value
+			}
+
+			m := value.Value.(map[FieldID]interface{})
+			for j, dt := range r.Schema() {
+				if r.Column(j).IsNull(i) {
+					m[j] = nil
+				} else {
+					d, ok := serdeMap[dt].deserialize(r.Column(j), i)
+					if ok {
+						m[j] = d // TODO: avoid memory copy here.
+					} else {
+						return errors.New(fmt.Sprintf("unexpected type %s", dt))
+					}
+				}
+			}
+
+			if _, ok := m[common.RowIDField]; !ok {
```

Comment:
```
rowID, ok := m[common.RowIDField].(int64)
if !ok {
    return fmt.Errorf("no row id column found")
}
value.ID = rowID
```
to avoid twice mapping